Add X10 devices as ISY994 switches

taylorsilva opened the pull request:
## Description:

This PR adds two things that I would like feedback on.
1. ~A new flag that adds X10 devices from the ISY994 as switches. I was not satisfied with having to write 20+ programs for all the X10 devices I have at my place. I think adding all X10 devices as switches is a decent middle ground instead of adding full support for them.~
Adds X10 devices as switches.
2. A warning message to help users debug why their ISY994 devices aren't being added as entities. I spent a lot of time trying to figure out why my X10 devices weren't being added and the current logging didn't provide enough detail. I had to dig into the component code to figure out why it wasn't being added. Having a logging message like this would have helped.

~**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io):** home-assistant/home-assistant.io#10271~

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

homeassistant commented:
Hi @taylorsilva, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#26342).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented on homeassistant/components/isy994/__init__.py:
```diff
@@ -386,6 +389,12 @@ def _categorize_weather(hass: HomeAssistant, climate) -> None:
 
 
+def _add_x10_filters():
+    global NODE_FILTERS
```
`global` is not allowed. A refactor may be needed.

taylorsilva replied:
I gave it a go at a refactor to avoid the usage of `global`. Thanks for the feedback Martin!

OverloadUT replied:
I don't believe this refactor was necessary. It seems to me that the only issue was the `global NODE_FILTERS` line, which could simply be removed. The `NODE_FILTERS` variable is defined up at the top, same as a ton of others, and is accessible to any functions here without needing to use `global`. You'll note that `NODE_FILTERS` is used in a whole bunch of functions.

MartinHjelmare commented:
@OverloadUT can you review?

taylorsilva commented:
As per @MartinHjelmare's comment, I got rid of the usage of `global` by doing a little refactoring. `node_filters` is now a dict that gets passed around between functions instead of something accessed globally.

Let me know what you think! @MartinHjelmare @OverloadUT

OverloadUT commented:
I think we should remove the config option and just have X10 devices be supported as switches always. There's not a lot of reason to treat X10 devices as any different than everything else, which means we include them, and the user can use the ignore feature if they want to remove any.

When I wrote the current method of translating ISY nodes in to Hass entities I did not exclude X10 on purpose; I simply don't have any and so I didn't think of it. :)

taylorsilva commented:
@OverloadUT I figured that was the case with X10's not being there lol 
Well I had fun hacking away at this, but I'll shrink this PR down to just directly adding X10's to the switches filter

OverloadUT commented:
Even if a lot of that work gets tossed away, you've still managed to get a pretty solid understanding of why that complicated code is all there and how it works, so you'll be able to contribute potential ISY changes in the future!

taylorsilva committed:
Add X10 devices as switches

Signed-off-by: Taylor Silva <taylorsilva@outlook.com>

taylorsilva committed:
Add logging message for when unsupported device is found.

This will help a user figure out why they aren't seeing all devices.

Signed-off-by: Taylor Silva <taylorsilva@outlook.com>

taylorsilva commented:
Updated the PR so it adds X10 devices as switches.

I also added a logging message that that lets a user know why their device wasn't added to HA. Not sure if this is the best place to put it, let me know what you think. This would've helped me figure out why my X10 devices weren't being added.

MartinHjelmare commented on homeassistant/components/isy994/__init__.py:
```diff
@@ -182,6 +182,7 @@ def _check_for_node_def(hass: HomeAssistant, node, single_domain: str = None) ->
             hass.data[ISY994_NODES][domain].append(node)
             return True
 
+    _LOGGER.warning("unsupported node: {node.name}, type: {node.type}")
```
```suggestion
    _LOGGER.warning("Unsupported node: %s, type: %s", node.name, node.type)
```

taylorsilva committed:
Update homeassistant/components/isy994/__init__.py

Co-Authored-By: Martin Hjelmare <marhje52@kth.se>

MartinHjelmare commented:
@OverloadUT ok?

OverloadUT commented on homeassistant/components/isy994/__init__.py:
```diff
@@ -114,7 +114,7 @@
     },
     "switch": {
-        "uom": ["2", "78"],
+        "uom": ["2", "78", "100"],
```
I'm a little hesitant about this one. 100 is a pretty generic uom in the ISY documentation.

I'm wondering if we can just go with the `insteon_type` and leave uom unchanged, as only one of these needs to match the node for the code to assign it to a domain.

Are you testing on 5.x or 4.x?

taylorsilva replied:
Testing on `5.x`. `5.0.15` to be specific. Omitting `100` works just fine.

```suggestion
        "uom": ["2", "78"],
```

taylorsilva committed:
Update homeassistant/components/isy994/__init__.py

100 is too generic of an ISY uom. We don't want to accidentally add non-switch devices.

OverloadUT commented:
@MartinHjelmare Good to go
